src/hotspot/share/gc/z/c2/zBarrierSetC2.cpp

```diff
@@ -296,6 +296,13 @@ void ZBarrierSetC2::clone_at_expansion(PhaseMacroExpand* phase, ArrayCopyNode* a
     Node* dest_offset = ac->in(ArrayCopyNode::DestPos);
     Node* length = ac->in(ArrayCopyNode::Length);
 
+    if (bt == T_OBJECT) {
+      // BarrierSetC2::arraycopy_payload_base_offset 8-byte aligns the offset.
+      // Make sure the offset points to the first element in the array when cloning
+      // object arrays. Otherwise, load barriers are applied to parts of the header.
+      src_offset = phase->longcon(arrayOopDesc::base_offset_in_bytes(T_OBJECT));
```

Comment:
Add comment that `BarrierSetC2::clone()` uses `arraycopy_payload_base_offset(is_array);` to set src and dst offsets. And add assert that `src_offset` and `dest_offset` are indeed equal to that constant.

My concern is that `CloneArray` could be set in other places in a future (currently only `BarrierSetC2::clone()` uses it) and offset could be different (may be already set correct values as you do).

Don't forget about RDP2 process.

Reply:
Thanks Vladimir. I've adjusted the comments and added asserts.

I'll request approval for integration into JDK 17.